File: Apps/AquaScriptCore/Core/VFS.cs
using System;
using System.Runtime.CompilerServices;

namespace AquaEngine
{
	public static class VFS
	{
		public static bool Exists(string path) => _Exists(path);
		public static string CurrentDirectory => _GetCurrentDirectory();
		public static string GetAbsolutePath(string path, bool requireFileExists = false, bool suppressWarnings = false) =>
			_GetAbsolutePath(path, requireFileExists, suppressWarnings);

		#region Mounting
		public static bool HasMount(string mountPoint) => _HasMount(mountPoint);

		public static void Mount(string mountPoint) => _Mount(mountPoint);
		public static void Mount(string mountPoint, string path) => _MountPath(mountPoint, path);

		public static void Unmount(string mountPoint) => _Unmount(mountPoint);
		public static void Unmount(string mountPoint, string path) => _UnmountPath(mountPoint, path);
		#endregion

		#region Read and Write
		public static byte[] Read(string path) => _Read(path);
		public static string ReadText(string path) => _ReadText(path);

		public static void Write(string path, byte[] data) => _Write(path, data);
		public static void Write(string path, string contents) => _WriteText(path, contents);
		#endregion

		#region File Operations
		public static void Remove(string path) => _Remove(path);
		public static void Copy(string source, string target) => _Copy(source, target);
		public static void Move(string source, string target) => _Move(source, target);
		#endregion

		#region Internal Calls
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern bool _Exists(string vfsPath);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern string _GetCurrentDirectory();
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern string _GetAbsolutePath(string vfsPath, bool requireFileExists, bool suppressWarnings);

		// Mount & Unmount
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern bool _HasMount(string path);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _Mount(string mountPoint);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _Unmount(string mountPoint);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _MountPath(string mountPoint, string path);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _UnmountPath(string mountPoint, string mountPath);

		// Read & Write
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern byte[] _Read(string path);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _Write(string path, byte[] content);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern string _ReadText(string path);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _WriteText(string path, string content);

		// File Operations
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _Remove(string path);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _Move(string source, string target);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _Copy(string source, string target);
		#endregion
	}
}

File: Engine/Source/Scripting/InternalCalls/VFS.cpp
#include <mono/jit/jit.h>
#include <spdlog/spdlog.h>
#include <AquaEngine/IO/VFS.hpp>
#include <AquaEngine/Scripting/Assembly.hpp>
#include <AquaEngine/Scripting/InternalCalls.hpp>

using namespace std;
using namespace AquaEngine::IO;

ADD_MANAGED_METHOD(VFS, GetCurrentDirectory, MonoString*, (), AquaEngine.IO)
{ return mono_string_new(mono_domain_get(), VFS::GetCurrentDirectory().string().c_str()); }

ADD_MANAGED_METHOD(VFS, HasMount, bool, (MonoString* path), AquaEngine.IO)
{ return VFS::HasMount(mono_string_to_utf8(path)); }

ADD_MANAGED_METHOD(VFS, Mount, void, (MonoString* mountPoint), AquaEngine.IO)
{ VFS::Mount(mono_string_to_utf8(mountPoint)); }

ADD_MANAGED_METHOD(VFS, MountPath, void, (MonoString* mountPoint, MonoString* mountPath), AquaEngine.IO)
{ VFS::Mount(mono_string_to_utf8(mountPoint), mono_string_to_utf8(mountPath)); }

ADD_MANAGED_METHOD(VFS, Unmount, void, (MonoString* mountPoint), AquaEngine.IO)
{ VFS::Unmount(mono_string_to_utf8(mountPoint)); }

ADD_MANAGED_METHOD(VFS, UnmountPath, void, (MonoString* mountPoint, MonoString* mountPath), AquaEngine.IO)
{ VFS::Unmount(mono_string_to_utf8(mountPoint), mono_string_to_utf8(mountPath)); }

ADD_MANAGED_METHOD(VFS, CreateDirectory, bool, (MonoString* path), AquaEngine.IO)
{ return VFS::CreateDirectory(mono_string_to_utf8(path)); }

ADD_MANAGED_METHOD(VFS, DeleteDirectory, bool, (MonoString* path), AquaEngine.IO)
{ return VFS::DeleteDirectory(mono_string_to_utf8(path)); }

ADD_MANAGED_METHOD(VFS, ReadText, MonoString*, (MonoString* path), AquaEngine.IO)
{
	string content = VFS::ReadText(mono_string_to_utf8(path));
	return mono_string_new(mono_domain_get(), content.c_str());
}

ADD_MANAGED_METHOD(VFS, WriteText, void, (MonoString* path, MonoString* content), AquaEngine.IO)
{ VFS::WriteText(mono_string_to_utf8(path), mono_string_to_utf8(content)); }

ADD_MANAGED_METHOD(VFS, Read, MonoArray*, (MonoString* path), AquaEngine.IO)
{
	vector<unsigned char> data = VFS::Read(mono_string_to_utf8(path));

	MonoArray* array = mono_array_new(mono_domain_get(), mono_get_byte_class(), data.size());
	for(size_t i = 0; i < data.size(); i++)
		mono_array_set(array, unsigned char, i, data[i]);
	return array;
}

ADD_MANAGED_METHOD(VFS, Write, void, (MonoString* path, MonoArray* array), AquaEngine.IO)
{
	size_t dataLength = mono_array_length(array);
	vector<unsigned char> data;
	data.reserve(dataLength);

	for(size_t i = 0; i < dataLength; i++)
		data.emplace_back(mono_array_get(array, unsigned char, i));

	VFS::Write(mono_string_to_utf8(path), data);
}

ADD_MANAGED_METHOD(VFS, Remove, void, (MonoString* path), AquaEngine.IO)
{ VFS::Remove(mono_string_to_utf8(path)); }

ADD_MANAGED_METHOD(VFS, Move, void, (MonoString* from, MonoString* to), AquaEngine.IO)
{ VFS::Move(mono_string_to_utf8(from), mono_string_to_utf8(to)); }

ADD_MANAGED_METHOD(VFS, Copy, void, (MonoString* from, MonoString* to), AquaEngine.IO)
{ VFS::Copy(mono_string_to_utf8(from), mono_string_to_utf8(to)); }

ADD_MANAGED_METHOD(VFS, GetAbsolutePath, MonoString*, (MonoString* path, bool requireFileExists, bool suppressWarnings), AquaEngine.IO)
{ return mono_string_new(mono_domain_get(), VFS::GetAbsolutePath(mono_string_to_utf8(path), requireFileExists, suppressWarnings).c_str()); }

ADD_MANAGED_METHOD(VFS, Exists, bool, (MonoString* path), AquaEngine.IO)
{ return VFS::Exists(mono_string_to_utf8(path)); }

ADD_MANAGED_METHOD(VFS, GetFiles, MonoArray*, (MonoString* directoryRaw, bool recursive), AquaEngine.IO)
{
	char* directory = mono_string_to_utf8(directoryRaw);
	vector<VFSFile> files = VFS::GetFiles(directory, recursive);

	MonoArray* output = mono_array_new(mono_domain_get(), mono_get_string_class(), (uintptr_t)files.size());
	for (size_t i = 0; i < files.size(); i++)
	{
		string path = files[i].FullPath();
		if (files[i].Directory)
			path += "/";
		mono_array_set(output, MonoString*, i, mono_string_new(mono_domain_get(), path.c_str()));
	}

	mono_free(directory);
	return output;
}